patina_dxe_core/src/image.rs

```diff
@@ -482,12 +522,132 @@ impl DxeCoreGlobalImageData {
         self.image_start_contexts = Vec::new();
     }
+
+    /// Finds the DXE Core memory allocation module HOB and uses it to produce the loaded image protocol.
+    fn install_dxe_core_image(&mut self, hob_list: &HobList, system_table: &mut EfiSystemTable) {
+        let dxe_core_hob = hob_list
+            .iter()
+            .find_map(|hob| {
+                if let Hob::MemoryAllocationModule(module) = hob
+                    && module.module_name == guids::DXE_CORE
+                {
+                    Some(module)
+                } else {
+                    None
+                }
+            })
+            .expect("Did not find MemoryAllocationModule Hob for DxeCore. Use patina::guid::DXE_CORE as FFS GUID.");
+
+        let mut image_info = empty_image_info();
+        image_info.system_table = system_table as *mut _ as *mut efi::SystemTable;
+        image_info.image_base = dxe_core_hob.alloc_descriptor.memory_base_address as *mut c_void;
+        image_info.image_size = dxe_core_hob.alloc_descriptor.memory_length;
+
+        // The entry point in the HOB is a u64 address. transmute it to the correct function pointer type.
+        // SAFETY: The module entry_point is spec defined as the below function signature.
+        let entry_point = unsafe {
+            transmute::<u64, extern "efiapi" fn(*mut c_void, *mut r_efi::system::SystemTable) -> r_efi::base::Status>(
+                dxe_core_hob.entry_point,
+            )
+        };
+
+        // SAFETY: The DXE Core HOB information is valid and points to a valid PE image. This ensures that the start
+        // address and length are accurate.
+        let dxe_core_image_buffer = unsafe {
+            from_raw_parts(
+                dxe_core_hob.alloc_descriptor.memory_base_address as *const u8,
+                dxe_core_hob.alloc_descriptor.memory_length as usize,
+            )
+        };
+
+        let pe_info = UefiPeInfo::parse(dxe_core_image_buffer).expect("Failed to parse PE info for DXE Core");
+
+        let private_image_data =
+            PrivateImageData::new_from_static_image(image_info, dxe_core_image_buffer, entry_point, &pe_info);
+
+        let handle = core_install_protocol_interface(
+            Some(protocol_db::DXE_CORE_HANDLE),
+            efi::protocols::loaded_image::PROTOCOL_GUID,
+            private_image_data.image_info.as_ref() as *const efi::protocols::loaded_image::Protocol as *mut c_void,
+        )
+        .unwrap_or_else(|err| panic!("Failed to install dxe core image handle: {err:?}"));
+
+        assert_eq!(handle, protocol_db::DXE_CORE_HANDLE);
+
+        initialize_debug_image_info_table(system_table);
+        core_new_debug_image_info_entry(
+            EfiDebugImageInfoNormal::EFI_DEBUG_IMAGE_INFO_TYPE_NORMAL,
+            private_image_data.image_info.as_ref() as *const efi::protocols::loaded_image::Protocol,
+            handle,
+        );
+
+        self.private_image_data.insert(handle, private_image_data);
```

Comment:
As a defensive measure, consider moving this up to right after core_install_protocol_interface. Once that succeeds, external agents can find a pointer to private_image_data, and so it needs to remain valid. This function is presently infallible but if that were to change in the future and an early exit were to occur between the install_protocol and the insertion of the private_image_data into the long-lived map, you could end up with a dangling interface. Putting them closer together is an aid to the programmer that they might be related :).